---
title: "Network analysis of IBD GWAS"
output: html_document
---

## Introduction
To explore the value of network propagation from GWAS of specific traits, we have defined 2 sets of seed genes for inflammatory bowel disease (IBD).

1. Curated: 37 selected high-confidence causal genes, from expert curation
2. L2G: 110 curated genes with L2G score > 0.5

We have run network propagation with these seed genes, and for each set we compare against 1,000 runs of network propagation with the same number of randomly selected genes as seed genes. From these runs we can define the Pagerank percentile of each gene relative to randomised inputs.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(GenomicRanges)
library(tidyverse)
library(ggExtra)
library(annotables)
library(cowplot)
options(stringsAsFactors = F)
knitr::opts_chunk$set(fig.width=8, fig.height=5)
theme_set(theme_bw())
saveFiles = T

root_dir = "/Users/jeremys/work/opentargets/gene_network/ibd"
analysis_dir = "/Users/jeremys/work/opentargets/gene_network/ibd/2021_05_06"

l2g.loci.df = read_tsv(file.path(analysis_dir, "l2g.IBD_loci.merged.annotated.tsv"))
l2g.genes.df = read_tsv(file.path(analysis_dir, "l2g.IBD_genes.merged.tsv"))

network1.df = read_csv(file.path(analysis_dir, "input/zsco_set1.csv.gz")) %>%
  mutate(pagerank_pctile = rankingIte1000.node / 10,
         seed_gene = padj > 0) %>%
  rename(gene_id = ENSG, geneSymbol = gene)

network2.df = read_csv(file.path(analysis_dir, "input/zsco_set2.csv.gz")) %>%
  mutate(pagerank_pctile = rankingIte1000.node / 10,
         seed_gene = padj > 0) %>%
  rename(gene_id = ENSG, geneSymbol = gene)

seed_genes_curated = network1.df %>% filter(seed_gene) %>% .$geneSymbol %>% unique()

seed_genes_l2g  = network2.df %>% filter(seed_gene) %>% .$geneSymbol %>% unique()
seed_genes_l2g = seed_genes_l2g[!seed_genes_l2g %in% seed_genes_curated]
```

The number of unique genes in the network should be the same for each of the seed gene selection sets:

```{r, warning=FALSE, message=FALSE, echo=FALSE}
network1.summary.df = network1.df %>%
  filter(!seed_gene) %>%
  group_by(gene_id) %>%
  summarise(numRuns = n(),
            Trait = if_else(numRuns == 1, first(Trait), ""),
            geneSymbol = dplyr::first(geneSymbol),
            pagerank.mean = mean(page.rank),
            pagerank = median(page.rank),
            degree = dplyr::first(degree),
            Zsco.pagerank.node.mean = mean(Zsco.page.rank.node),
            Zsco.pagerank.node = median(Zsco.page.rank.node),
            pagerank_pctile.mean = mean(pagerank_pctile),
            pagerank_pctile = median(pagerank_pctile),
            seed_gene = (numRuns == 1))

network2.summary.df = network2.df %>%
  filter(!seed_gene) %>%
  group_by(gene_id) %>%
  summarise(numRuns = n(),
            Trait = if_else(numRuns == 1, first(Trait), ""),
            geneSymbol = dplyr::first(geneSymbol),
            pagerank.mean = mean(page.rank),
            pagerank = median(page.rank),
            degree = dplyr::first(degree),
            Zsco.pagerank.node.mean = mean(Zsco.page.rank.node),
            Zsco.pagerank.node = median(Zsco.page.rank.node),
            pagerank_pctile.mean = mean(pagerank_pctile),
            pagerank_pctile = median(pagerank_pctile),
            seed_gene = (numRuns == 1))

if (saveFiles) {
  write_tsv(network1.summary.df %>% arrange(desc(pagerank_pctile)), file.path(analysis_dir, "network.curated.summary.tsv"), na = "")
  write_tsv(network2.summary.df %>% arrange(desc(pagerank_pctile)), file.path(analysis_dir, "network.l2g.summary.tsv"), na = "")
}
```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
#length(unique(network1.df$gene_id))
nrow(network1.summary.df)
nrow(network2.summary.df)

# idx = which(!network1.summary.df$gene_id %in% network2.summary.df$gene_id)
# View(network1.summary.df[idx,])
# "ENSG00000075426" %in% network2.df$gene_id
# "FOSL2" %in% network2.df$geneSymbol
# "ENSG00000075426" %in% network2.df$gene_id
# "FOSL2" %in% network2.summary.df$geneSymbol

#length(unique(network1.df %>% filter(padj > 0) %>% .$gene_id))
#length(unique(network2.df %>% filter(padj > 0) %>% .$gene_id))
#length(unique(network3.df %>% filter(padj > 0) %>% .$gene_id))
```


## Distributions

We first consider the network scores based on the stringent set of 33 curated IBD genes.

### Distribution of Pagerank percentile

```{r, warning=FALSE, message=FALSE, echo=FALSE}
#ggplot(network1.df, aes(x=pagerank_pctile)) + geom_histogram(bins = 100)
ggplot(network1.summary.df, aes(x=pagerank_pctile)) + geom_histogram(bins = 100) + ggtitle("Pagerank pctile distribution (curated seed gene network)")
ggplot(network2.summary.df, aes(x=pagerank_pctile)) + geom_histogram(bins = 100) + ggtitle("Pagerank pctile distribution (L2G seed gene network)")
```

Each of the network sets has a peak of occurrence at a particular (fairly low) pagerank percentile, which differs between the networks. There is also some depletion of very low pagerank percentiles for the stringent seed gene networks, but apart from that, the distribution of pagerank percentile is fairly uniform.

Let's look at a QQ plot to see how the quantiles of this distribution relate to the uniform distribution.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
ggplot(network1.summary.df, aes(x=pagerank_pctile)) + stat_ecdf(geom = "step") +
  geom_abline(slope = 0.01, intercept = 0, col = "blue") +
  ylab("Distribution quantile") + ggtitle("Pagerank pctile QQ plot (curated seed gene network)")

ggplot(network2.summary.df, aes(x=pagerank_pctile)) + stat_ecdf(geom = "step") +
  geom_abline(slope = 0.01, intercept = 0, col = "blue") +
  ylab("Distribution quantile") + ggtitle("Pagerank pctile QQ plot (L2G seed gene network)")
```

These are close enough to uniform (better than for AD) that we can use the Pagerank percentiles as-is. (Especially since we're not concerned about the very low range of Pagerank anyway.) So in general, a pagerank percentile of 80% means (to a close approximation) that the gene has a network score greater than 80% of genes.

### Distribution of node degree

```{r, warning=FALSE, message=FALSE, echo=FALSE}
ggplot(network1.summary.df, aes(x=degree, col=seed_gene, fill=seed_gene)) +
  geom_density(alpha = 0.3, size = 1) +
  scale_x_log10() +
  ggtitle("Gene degree by gene category (curated seed gene network)")

ggplot(network2.summary.df, aes(x=degree, col=seed_gene, fill=seed_gene)) +
  geom_density(alpha = 0.3, size = 1) +
  scale_x_log10() +
  ggtitle("Gene degree by gene category (L2G seed gene network)")
```

Summary for seed genes (for the 3 networks):

```{r, warning=FALSE, message=FALSE}
network1.summary.df %>% filter(seed_gene) %>% .$degree %>% summary()
network2.summary.df %>% filter(seed_gene) %>% .$degree %>% summary()
```

Summary for non-seed genes:

```{r, warning=FALSE, message=FALSE}
network1.summary.df %>% filter(!seed_gene) %>% .$degree %>% summary()
```

Our seed genes have a higher node degree on average than remaining genes. The L2G-based seed genes tend to have lower degree than the curated genes, but still higher than the average gene.



```{r, warning=FALSE, message=FALSE, echo=FALSE}
# A function to split a continuous annotation into categories
categorize = function(vals, thresholds) {
  threshold = sort(thresholds)
  if (length(thresholds) < 1) {
    return(NULL)
  }
  catLevels = c(paste0("<= ", threshold[1]))
  cats = rep(catLevels[1], length(vals))
  for (i in 1:length(thresholds)) {
    if (i == length(thresholds)) {
      catLevels[i+1] = paste0("> ", threshold[i])
      cats[vals > threshold[i]] = catLevels[i+1]
    } else {
      catLevels[i+1] = paste0(threshold[i], "-", threshold[i+1])
      cats[vals > threshold[i]] = catLevels[i+1]
    }
  }
  cats = factor(cats, levels = catLevels)
  return(cats)
}

stratifiedOddsRatios = function(df, cat1, cat2, vsBaseCategory = T) {
  df = df[, c(cat1, cat2)]
  if (any(is.na(df))) {
    warning("stratifiedOddsRatios: Note - input table has NA values. These will be omitted.")
    df = na.omit(df)
  }
  cat1levels = levels(pull(df,cat1))
  cat2levels = levels(pull(df,cat2))
  cat1base = levels(pull(df,cat1))[1]
  cat2base = levels(pull(df,cat2))[1]
  oddsRatio.df = data.frame(comparisonStr = character(), cat1 = character(), cat2 = character(),
                            estimate = numeric(), confint_lo = numeric(), confint_hi = numeric())
  cat1higherLevels = cat1levels[cat1levels != cat1base]
  cat2higherLevels = cat2levels[cat2levels != cat2base]
  
  for (cat1level in cat1higherLevels) {
    for (cat2level in cat2higherLevels) {
      if (vsBaseCategory) {
        mat = matrix(c(sum(df[,cat1] == cat1level & df[,cat2] == cat2level),
                       sum(df[,cat1] == cat1base & df[,cat2] == cat2level),
                       sum(df[,cat1] == cat1level & df[,cat2] == cat2base),
                       sum(df[,cat1] == cat1base & df[,cat2] == cat2base)), nrow=2)
        comparisonStr = sprintf("%s vs. %s to %s vs. %s", cat1level, cat1base, cat2level, cat2base)
      } else {
        mat = matrix(c(sum(df[,cat1] == cat1level & df[,cat2] == cat2level),
                       sum(df[,cat1] != cat1level & df[,cat2] == cat2level),
                       sum(df[,cat1] == cat1level & df[,cat2] != cat2level),
                       sum(df[,cat1] != cat1level & df[,cat2] != cat2level)), nrow=2)
        comparisonStr = sprintf("%s to %s", cat1level, cat2level)
      }
      res = fisher.test(mat)
      oddsRatio.df = bind_rows(oddsRatio.df,
                               data.frame(comparisonStr, cat1 = cat1level, cat2 = cat2level,
                                          estimate = res$estimate, confint_lo = res$conf.int[1], confint_hi = res$conf.int[2]))
    }
  }
  oddsRatio.df
}

stratifiedOddsRatioPlot = function(or.df, cat1name, cat2name, cat1levels = NULL, cat2levels = NULL) {
  if (!is.null(cat1levels)) {
    or.df$cat1 = factor(or.df$cat1, levels = cat1levels)
  }
  if (!is.null(cat2levels)) {
    or.df$cat2 = factor(or.df$cat2, levels = cat2levels)
  }
  ggplot(or.df, aes(x=cat1, y=estimate, fill=cat2)) +
    geom_bar(stat = "identity", position = "dodge") +
    geom_errorbar(aes(ymin = confint_lo, ymax = confint_hi), position = position_dodge2(width = 0.4, padding = 0.6), color = "grey40") +
    scale_y_log10() + ylab("Odds ratio") +
    xlab(cat1name) +
    scale_fill_discrete(name = cat2name)
}
```


## Network scores of seed genes

We assess the extent to which our seed genes have higher Pagerank percentiles than average.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
#, fig.width=5, fig.height=4
l2g.net.df = l2g.genes.df %>% left_join(network1.df, by="gene_id")

l2g.net.uniq.df = l2g.net.df %>%
  filter(!duplicated(gene_id), !is.na(pagerank_pctile))

network1.summary.df = network1.summary.df %>%
  mutate(locus_gene = gene_id %in% l2g.genes.df$gene_id,
         gene_type = if_else(geneSymbol %in% seed_genes_curated,
                             "curated gene",
                             if_else(geneSymbol %in% seed_genes_l2g,
                                     "L2G gene",
                                     if_else(locus_gene, "locus gene", "other gene")))) %>%
  mutate(gene_type = factor(as.character(gene_type), levels = c("other gene", "locus gene", "L2G gene", "curated gene")))


p = ggplot(network1.summary.df, aes(x = gene_type, y = pagerank_pctile, fill = gene_type)) +
  geom_jitter(width = 0.1, alpha = 0.2) +
  geom_violin(alpha = 0.3) +
  geom_boxplot(outlier.shape = NA, width = 0.07, alpha = 0.8, fill="white") +
  theme_bw(12) +
  ggtitle("Pagerank pctile by gene group (curated seed gene network)") +
  scale_fill_discrete(guide=F) +
  ylab("Pagerank percentile") +
  theme(axis.text.x = element_text(angle = 25, hjust = 1), axis.title.x = element_blank())
print(p)

if (saveFiles) {
  pdf(file.path(analysis_dir, "pagerank_pctile_violin.pdf"), width=4.7, height=3.5)
  print(p + theme(plot.title = element_blank()))
  dev.off()
}
summary(network1.summary.df %>% filter(seed_gene) %>% .$pagerank_pctile)
#network1.summary.df %>% filter(gene_type == "curated gene") %>% nrow()

wilcox.test(network1.summary.df %>% filter(seed_gene) %>% .$pagerank_pctile,
            network1.summary.df %>% filter(gene_type == "locus gene") %>% .$pagerank_pctile,
            alternative = "greater")
# ks.test(network1.summary.df %>% filter(seed_gene) %>% .$pagerank_pctile,
#             network1.summary.df %>% filter(gene_type == "locus gene") %>% .$pagerank_pctile,
#             alternative = "less")
# t.test(network1.summary.df %>% filter(seed_gene) %>% .$pagerank_pctile,
#        network1.summary.df %>% filter(gene_type == "locus gene") %>% .$pagerank_pctile,
#        alternative = "greater")

```

They are much higher on average, with a few outliers.

Let's also look at this for the L2G network.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
network2.summary.df = network2.summary.df %>%
  mutate(locus_gene = gene_id %in% l2g.genes.df$gene_id,
         gene_type = if_else(geneSymbol %in% seed_genes_curated,
                             "curated gene",
                             if_else(geneSymbol %in% seed_genes_l2g,
                                     "L2G gene",
                                     if_else(locus_gene, "locus gene", "other gene")))) %>%
  mutate(gene_type = factor(as.character(gene_type), levels = c("other gene", "locus gene", "L2G gene", "curated gene")))

ggplot(network2.summary.df, aes(x = gene_type, y = pagerank_pctile, fill = gene_type)) +
  geom_jitter(width = 0.1, alpha = 0.2) +
  geom_violin(alpha = 0.3) +
  geom_boxplot(outlier.shape = NA, width = 0.07, alpha = 0.8, fill="white") +
  ggtitle("Pagerank pctile QQ plot (L2G seed gene network)")

summary(network2.summary.df %>% filter(seed_gene) %>% .$pagerank_pctile)
#network2.summary.df %>% filter(gene_type == "L2G gene") %>% nrow()

wilcox.test(network2.summary.df %>% filter(gene_type == "L2G gene") %>% .$pagerank_pctile,
            network2.summary.df %>% filter(gene_type == "locus gene") %>% .$pagerank_pctile,
            alternative = "greater")


```

In the network based on 110 L2G > 0.5 genes, the curated genes have slightly lower network scores than before, whereas the L2G genes are slightly higher. I'm not sure what this indicates - it could be interesting? Perhaps the curated genes are more biased towards specific coherent (known) pathways, and hence being selected as a curated gene means you are likely to be in those (high-network scoring) pathways... whereas L2G genes are less biased, as so do better on average in the less biased network?

Let's see the correlation between scores for the 2 networks.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
merged_network = network1.summary.df %>%
  rename(curated_PRpctile = pagerank_pctile) %>%
  left_join(network2.summary.df %>% rename(L2G_PRpctile = pagerank_pctile) %>% select(gene_id, L2G_PRpctile), by="gene_id") %>%
  left_join(l2g.genes.df %>% select(gene_id, geneSymbol = symbol, l2g_score_mean), by="geneSymbol")

ggplot(merged_network, aes(x=curated_PRpctile, y=L2G_PRpctile)) +
  geom_point(alpha = 0.1)
cor.test(merged_network$curated_PRpctile, merged_network$L2G_PRpctile, method = "spearman")
```


## Network scores vs. locus2gene

Let's look at the correlation (scatterplot) of locus2gene score and network score with each as the x-axis variable in turn.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
ggplot(l2g.net.uniq.df, aes(x=l2g_score_mean, y=pagerank_pctile)) +
  geom_point(alpha=0.4) +
  geom_smooth() + ylab("Curated network PRpctile")

ggplot(l2g.net.uniq.df, aes(y=l2g_score_mean, x=pagerank_pctile)) +
  geom_point(alpha=0.4) +
  geom_smooth() + xlab("Curated network PRpctile")
```

A violin plot might give a clearer view of what network score to expect for highly confident genes (based on high l2g score). This is similar to just looking at the curated, strict, or lenient gene sets as we did above.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
l2g.net.uniq.df$pagerank_pctile.cat = categorize(l2g.net.uniq.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90, 95, 98))
l2g.net.uniq.df$l2g_score_mean.cat = categorize(l2g.net.uniq.df$l2g_score_mean, thresholds = c(0.2, 0.4, 0.6, 0.8))
ggplot(l2g.net.uniq.df, aes(x=l2g_score_mean.cat, y=pagerank_pctile)) +
  geom_violin(alpha = 0.9) +
  geom_jitter(width = 0.25, alpha = 0.2) +
  geom_boxplot(width = 0.05, outlier.shape = NA, fill="white") +
  ylab("Curated network PRpctile")
```

And similarly, but with axes reversed.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
ggplot(l2g.net.uniq.df, aes(x=pagerank_pctile.cat, y=l2g_score_mean)) +
  geom_violin(alpha = 0.9) +
  geom_boxplot(width = 0.05, outlier.shape = NA, fill="white") +
  xlab("Curated network PRpctile")
```


## Network scores per locus

We can look at these loci in more detail.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=8, fig.height=4.2}
network1_seedgenes = network1.summary.df %>% filter(seed_gene) %>% .$gene_id %>% unique()
network1_seedgene_loci = l2g.genes.df %>% filter(gene_id %in% network1_seedgenes) %>% .$l2g_locus_merged %>% unique()

l2g.seed_locus.df = l2g.genes.df %>% filter(l2g_locus_merged %in% network1_seedgene_loci)
#ViewDups(l2g.seed_locus.df, 6)
l2g.seed_locus.df = l2g.seed_locus.df %>%
  left_join(network1.summary.df, by = "gene_id") %>%
  group_by(l2g_locus_merged) %>%
  arrange(desc(seed_gene)) %>%
  mutate(locus_name = first(geneSymbol),
         has_seed_gene = any(seed_gene)) %>%
  select(l2g_locus_merged, seed_gene, geneSymbol, locus_name, everything()) %>% na.omit()

p = ggplot(l2g.seed_locus.df %>% filter(!seed_gene), aes(x=locus_name, y=pagerank_pctile, col=seed_gene, size=seed_gene)) +
  geom_point(alpha=0.5) +
  geom_point(data = l2g.seed_locus.df %>% filter(seed_gene), alpha=0.7) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("PageRank percentile per locus") +
  ylab("Pagerank percentile") + xlab("Seed gene") +
  scale_color_manual(name = "Seed gene", values = c("FALSE"="grey20", "TRUE"="blue")) +
  scale_size_manual(name = "Seed gene", values=c("TRUE"=2.5, "FALSE"=1.5))
print(p)

if (saveFiles) {
  pdf(file.path(analysis_dir, "pagerank_pct_per_locus.pdf"), width=8, height=4.2)
  print(p)
  dev.off()
}
```

(Caveat: this focuses on genes identified by Locus2Gene as being near Katie's GWAS loci. It's not a good definition of "locus" yet, which is why you have some loci here with no seed genes. Selection of locus genes should be made more robust in future.)

Dig into loci where our seed gene isn't among the top few.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=8, fig.height=6}
l2g.seed_locus.df = l2g.seed_locus.df %>%
  filter(has_seed_gene) %>%
  group_by(locus_name) %>%
  mutate(locus_rank = rank(-pagerank_pctile))

#View(l2g.seed_locus.df %>% arrange(desc(seed_gene), locus_rank) %>% select(locus_name, seed_gene, locus_rank, pagerank_pctile, everything()))

plot.df = l2g.seed_locus.df %>%
  filter(locus_name %in% (l2g.seed_locus.df %>% filter(seed_gene, locus_rank >= 3, pagerank_pctile < 90) %>% .$locus_name)) %>%
  group_by(locus_name) %>%
  mutate(locus_seed_gene_rank = max(if_else(seed_gene, locus_rank, 0)),
         label = if_else(locus_rank <= locus_seed_gene_rank, geneSymbol, ""))
#View(plot.df %>% select(locus_name, seed_gene, locus_rank, pagerank_pctile, geneSymbol, label, y_proba_full_model, everything()) %>% arrange(desc(locus_name), desc(y_proba_full_model)))

ggplot(plot.df %>% filter(!seed_gene),
       aes(x=locus_name, y=pagerank_pctile, col=seed_gene, size=seed_gene)) +
  geom_point(alpha=0.5) +
  geom_point(data = plot.df %>% filter(seed_gene), alpha=0.7) +
  geom_text(data = plot.df, aes(label = label), hjust = 0, nudge_x = 0.1, size = 2) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("Loci with genes having higher network score than seed genes") +
  ylab("Pagerank percentile") + xlab("Seed gene") +
  scale_color_manual(name = "Seed gene", values = c("FALSE"="grey20", "TRUE"="blue")) +
  scale_size_manual(name = "Seed gene", values=c("TRUE"=2.5, "FALSE"=1.5))


```

We can look at a scatterplot of individual loci to help determine which loci might have better candidates (based on network score and/or locus2gene) than the ones that were picked. Two examples are below, and plots for all loci are saved in a PDF.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=6, fig.height=4}
l2g_scatterplot = function(l2g.seed_locus.df, cur_locus_name) {
  p.df = l2g.seed_locus.df %>% filter(locus_name == cur_locus_name)
  ggplot(p.df, aes(x=l2g_score_mean, y=pagerank_pctile, col=seed_gene, size=seed_gene)) +
    geom_point(data = p.df %>% filter(!seed_gene), alpha=0.5) +
    geom_text(data = p.df %>% filter(!seed_gene), aes(label = geneSymbol), hjust = 0, nudge_x = 0.01, size = 2.0) +
    geom_point(data = p.df %>% filter(seed_gene), alpha=0.8) +
    geom_text(data = p.df %>% filter(seed_gene), aes(label = geneSymbol), hjust = 0, nudge_x = 0.01, size = 2.3) +
    ggtitle(cur_locus_name) +
    ylab("Pagerank percentile") + xlab("L2G score") +
    scale_color_manual(name = "Seed gene", values = c("FALSE"="grey20", "TRUE"="blue"), guide=F) +
    scale_size_manual(name = "Seed gene", values=c("TRUE"=2.5, "FALSE"=1.5), guide=F) +
    xlim(0, max(p.df$l2g_score_mean) + 0.1)
#    ggtitle(sprintf("Locus scatterplot: %s", cur_locus_name)) +
}

if (saveFiles) {
  locus_names = unique(l2g.seed_locus.df$locus_name)
  pdf(file = file.path(analysis_dir, "curated_network.locus_scatterplots.pdf"), width=7, height=5)
  for (cur_locus in locus_names) {
    print( l2g_scatterplot(l2g.seed_locus.df, cur_locus) )
  }
  dev.off()
}
print( l2g_scatterplot(l2g.seed_locus.df, "INAVA") )
print( l2g_scatterplot(l2g.seed_locus.df, "LRRK2") )
print( l2g_scatterplot(l2g.seed_locus.df, "BANK1") )

```

In most cases, for the loci where the curated gene has a low network score, it has a very high L2G, and no other genes do. This suggests to me that at these loci the network is missing something relevant.
One outlier is BANK1, which has high network score and L2G score, whereas SLC39A8 is selected, and is relatively low for both scores.


```{r, warning=FALSE, message=FALSE, echo=FALSE, eval=FALSE}
# Get genes in the lenient network which are in the same module as LACC1
lacc1_module = network2.df %>% filter(Trait == "AL512506.3", gene_id == "ENSG00000179630") %>% .$cluster.walktrap
lacc1_module_genes = network2.df %>% filter(Trait == "LACC1", cluster.walktrap == lacc1_module) %>% .$geneSymbol
# Get the intersection of genes in this module with our lenient seed genes
lacc1_module_seed_genes = lacc1_module_genes[lacc1_module_genes %in% seed_genes_lenient]
network3.df %>% filter(Trait == "SOX4", geneSymbol == "SOX4")
network3.df %>% filter(Trait == "ZEB2", geneSymbol == "ZEB2")
network3.df %>% filter(Trait == "SATB1", geneSymbol == "SATB1")

```


Notes:

* For most of these loci the seed gene is still near the top.
* SIRPG could be interesting at the FKBP1A locus (high L2G + network score)
* Both MST1 and APEH could be interesting at the MST1 locus
* For TNFSF8, I would put money on TNFSF15 being the gene, since it has a high L2G score (nearest gene), stronger coloc evidence (in the genetics portal at least), and higher network score.
* For ERAP2/ERAP1, I would be cautious in making any conclusions, since these have fairly low degree (number of connections), so the difference could be down to one connection - e.g. if one is missing for ERAP2 due to incompleteness in the network.

Let's look at this for all loci, based on the curated seed gene network.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=15, fig.height=6}
l2g.genes.net.df = l2g.genes.df %>%
  left_join(network1.summary.df, by = "gene_id") %>%
  group_by(l2g_locus_merged) %>%
  arrange(desc(seed_gene)) %>%
  mutate(locus_name = first(geneSymbol),
         has_seed_gene = any(seed_gene)) %>%
  select(l2g_locus_merged, seed_gene, geneSymbol, locus_name, gene_type, everything()) %>% na.omit()

#length(unique(l2g.genes.net.df %>% filter(chr <= 4) %>% .$l2g_locus_merged))

ggplot(l2g.genes.net.df %>% filter(gene_type == "locus gene"), aes(x=l2g_locus_merged, y=pagerank_pctile, col=gene_type, size=gene_type)) +
  geom_point(alpha=0.5) +
  geom_point(data = l2g.genes.net.df %>% filter(gene_type != "locus gene"), alpha=0.7) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("PageRank percentile per locus") +
  ylab("Pagerank percentile") + xlab("Locus") +
  scale_color_manual(name = "Gene group", values = c("locus gene"="grey20", "curated gene"="blue", "L2G gene"="red")) +
  scale_size_manual(name = "Gene group", values=c("locus gene"=1.5, "curated gene"=2.5, "L2G gene"=2.5))

```


```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=6, fig.height=4}
# Let's also save some l2g scatterplots for all loci based on the L2G >0.5 seed gene networks
network2_seedgenes = network2.summary.df %>% filter(seed_gene) %>% .$gene_id %>% unique()
network2_seedgene_loci = l2g.genes.df %>% filter(gene_id %in% network2_seedgenes) %>% .$l2g_locus_merged %>% unique()
l2g.locus.df = l2g.genes.df %>% filter(l2g_locus_merged %in% network2_seedgene_loci)
#ViewDups(l2g.seed_locus.df, 6)
l2g.locus.df = l2g.locus.df %>%
  left_join(network2.summary.df, by = "gene_id") %>%
  group_by(l2g_locus_merged) %>%
  arrange(desc(seed_gene), desc(l2g_score_mean)) %>%
  mutate(locus_name = first(geneSymbol),
         has_seed_gene = any(seed_gene)) %>%
  select(l2g_locus_merged, seed_gene, geneSymbol, locus_name, everything()) %>% na.omit()

if (saveFiles) {
  locus_names = unique(l2g.locus.df$locus_name)
  pdf(file = file.path(analysis_dir, "l2g_network.locus_scatterplots.pdf"), width=7, height=5)
  for (cur_locus in locus_names) {
    print( l2g_scatterplot(l2g.locus.df, cur_locus) )
  }
  dev.off()
}
#print( l2g_scatterplot(l2g.locus.df, "ICAM1") )
#print( l2g_scatterplot(l2g.locus.df, "LACC1") )

```


```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=10, fig.height=4}
# Save a file that has 3 locus scatter plots side by side for each locus,
# one for each of the networks
if (saveFiles) {
  l2g.network.df = l2g.genes.df %>%
    left_join(network2.summary.df %>% select(gene_id, seed_gene2 = seed_gene, prpct2 = pagerank_pctile, geneSymbol), by = "gene_id") %>%
    left_join(network1.summary.df %>% select(gene_id, seed_gene1 = seed_gene, prpct1 = pagerank_pctile), by = "gene_id") %>%
    group_by(l2g_locus_merged) %>%
    arrange(desc(seed_gene2), desc(l2g_score_mean)) %>%
    mutate(locus_name = first(geneSymbol),
           has_seed_gene = any(seed_gene2 | seed_gene1)) %>%
    select(l2g_locus_merged, locus_name, gene_id, geneSymbol, seed_gene2, seed_gene1, starts_with("l2g"), everything()) %>%
    filter(!is.na(prpct2))
  
  locus_names = unique(l2g.network.df$locus_name)
  l2g.network1.df = l2g.genes.df %>%
    left_join(network1.summary.df %>% select(gene_id, geneSymbol, seed_gene, pagerank_pctile), by = "gene_id") %>% mutate(network="curated")
  l2g.network2.df = l2g.genes.df %>%
    left_join(network2.summary.df %>% select(gene_id, geneSymbol, seed_gene, pagerank_pctile), by = "gene_id") %>% mutate(network="l2g")
  l2g.network.df = bind_rows(l2g.network1.df, l2g.network2.df) %>%
    group_by(l2g_locus_merged) %>%
    arrange(desc(seed_gene), desc(l2g_score_mean)) %>%
    mutate(locus_name = first(geneSymbol)) %>%
    ungroup() %>%
    mutate(network = factor(as.character(network), levels=c("curated", "l2g")))
  locus_names = unique(l2g.network.df$locus_name)

  pdf(file = file.path(analysis_dir, "locus_scatterplots.2networks.pdf"), width=10, height=4)
  for (cur_locus in locus_names) {
    p = l2g_scatterplot(l2g.network.df, cur_locus) + facet_wrap(~network, nrow = 1) + theme_bw(10)
    print(p)
  }
  dev.off()
}
```


## Pathway enrichments

Look at pathway enrichments for the top 100 or 1000 genes in the curated seed gene network. These are visualised below using GOSummaries, and the gProfiler outputs are saved as .tsv files.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=9, eval=FALSE}
library(GOsummaries)
library(gProfileR)

top100Genes_PRpctile = network1.summary.df %>% arrange(desc(pagerank_pctile)) %>% .$gene_id %>% .[1:100]
top1000Genes_PRpctile = network1.summary.df %>% arrange(desc(pagerank_pctile)) %>% .$gene_id %>% .[1:1000]

gs = gosummaries(list(top100PRpctile = top100Genes_PRpctile), custom_bg = network1.summary.df$gene_id)
plot(gs, fontsize = 11)

gs = gosummaries(list(top1000PRpctile = top1000Genes_PRpctile), custom_bg = network1.summary.df$gene_id)
plot(gs, fontsize = 11)

gp.top100 = gprofiler(query = top100Genes_PRpctile, custom_bg = network1.summary.df$gene_id, organism = "hsapiens")
gp.top1000 = gprofiler(query = top1000Genes_PRpctile, custom_bg = network1.summary.df$gene_id, organism = "hsapiens")
gp.top100 %>% arrange(p.value) %>% write_tsv(file.path(analysis_dir, "curated.top100.gprofiler.tsv"))
gp.top1000 %>% arrange(p.value) %>% write_tsv(file.path(analysis_dir, "curated.top1000.gprofiler.tsv"))

```


## IBD p-value network enrichment (10 kb window)

First look at the distribution of minp values across genes. (For Katie's IBD GWAS.)

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=6, fig.height=4}
vecQuantiles = function(x) {
  sapply(x, function(val) if (is.na(val)) { NA } else { (sum(x <= val, na.rm = T) - 1) / (sum(!is.na(x)) - 1) })
}

gene_minp_file = file.path(root_dir, "genes.pc.10kb_window.katie_ibd.minp.tsv.gz")

if (!file.exists(gene_minp_file)) {
  genes.minp.df = readr::read_tsv(file.path(root_dir, "genes.pc.10kb_window.katie_ibd.snp_overlaps.tsv.gz")) %>%
    filter(pvalue != -1) %>%
    na.omit() %>%
    group_by(gene_id) %>%
    summarise(chr = first(chr),
              gene_window_start = first(gene_window_start),
              gene_window_end = first(gene_window_end),
              snp_pos = first(snp_pos),
              snp_id = first(snp_id),
              minp = min(pvalue)) %>%
    filter(minp > -1) %>%
    ungroup() %>%
    mutate(minp.quantile = vecQuantiles(-minp))
  write_tsv(genes.minp.df, gene_minp_file)
} 
genes.minp.df = readr::read_tsv(gene_minp_file)

network1.minp.df = network1.summary.df %>%
  inner_join(genes.minp.df, by="gene_id")

ggplot(network1.minp.df %>% filter(minp > 1e-8), aes(x=-log10(minp))) +
  geom_histogram(bins=100) + ggtitle("Distribution of minp values (Katie IBD GWAS)")
#sum(is.na(network1.minp.df$minp))
#sum(network1.minp.df$minp < 5e-8, na.rm=T)
#quantile(network1.minp.df$minp, probs=seq(0, 1, 0.01), na.rm=T)

```

Do similarly for Jimmy's IBD GWAS.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=6, fig.height=4}
jimmy_gene_minp_file = file.path(root_dir, "genes.pc.10kb_window.jimmy_ibd.minp.tsv.gz")

if (!file.exists(jimmy_gene_minp_file)) {
  genes.jimmy.minp.df = readr::read_tsv(file.path(root_dir, "genes.pc.10kb_window.jimmy_ibd.snp_overlaps.tsv.gz")) %>%
    filter(pvalue != -1) %>%
    na.omit() %>%
    group_by(gene_id) %>%
    summarise(chr = first(chr),
              gene_window_start = first(gene_window_start),
              gene_window_end = first(gene_window_end),
              snp_pos = first(snp_pos),
              snp_id = first(snp_id),
              minp = min(pvalue)) %>%
    filter(minp > -1) %>%
    ungroup() %>%
    mutate(minp.quantile = vecQuantiles(-minp))
  write_tsv(genes.jimmy.minp.df, jimmy_gene_minp_file)
} 
genes.jimmy.minp.df = readr::read_tsv(jimmy_gene_minp_file)

network1.jimmy.minp.df = network1.summary.df %>%
  inner_join(genes.jimmy.minp.df, by="gene_id")

ggplot(network1.jimmy.minp.df %>% filter(minp > 1e-8), aes(x=-log10(minp))) +
  geom_histogram(bins=100) + ggtitle("Distribution of minp values (Jimmy IBD GWAS)")
#sum(is.na(network1.minp.df$minp))
#sum(network1.minp.df$minp < 5e-8, na.rm=T)
#quantile(network1.minp.df$minp, probs=seq(0, 1, 0.01), na.rm=T)

```

Jimmy's GWAS had only 157k variants. As you might expect, minp values are higher on average, since we're taking the minimum across fewer variants.

Now we look at the odds ratio for genes at a given network percentile to have low p-values (within specific ranges, relative to p > 0.01 and pagerank pctile < 50).

```{r, warning=FALSE, message=FALSE, echo=FALSE}
#, fig.width=5.5, fig.height=4
# network1.minp.df$pagerank_pctile.cat = categorize(network1.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90, 95))
# network1.minp.df$pagerank_pctile.cat = factor(network1.minp.df$pagerank_pctile.cat, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))
# 
# ggplot(network1.minp.df, aes(x=pagerank_pctile.cat, y=minp.quantile)) +
#   geom_boxplot() + geom_jitter(width=0.25, alpha=0.2)

network1.minp.df$minp_cat = categorize(network1.minp.df$minp, thresholds = c(0.01, 1e-4, 1e-6, 1e-8))
minp_cat_levels = c("> 0.01", "1e-04-0.01", "1e-06-1e-04", "1e-08-1e-06", "<= 1e-08")
network1.minp.df$minp_cat = factor(network1.minp.df$minp_cat, levels = minp_cat_levels)

network1.minp.df$pagerank_pctile.cat = categorize(network1.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90))
pagerank_pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "> 90")
network1.minp.df$pagerank_pctile.cat = categorize(network1.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90, 95, 98))
pagerank_pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "95-98", "> 98")
network1.minp.df$pagerank_pctile.cat = factor(network1.minp.df$pagerank_pctile.cat, levels = pagerank_pct_cat_levels)
#xtabs(~ minp_cat + pagerank_pctile.cat, data = network1.minp.df)
or.df = stratifiedOddsRatios(network1.minp.df, "minp_cat", "pagerank_pctile.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
#or.df$cat1[or.df$cat1 == "1e-06-1e-04"] = parse(text="10^-6 - 10^-4")
or.df$cat2 = factor(or.df$cat2, levels = pagerank_pct_cat_levels)

minp_cat_display_levels = c("> 0.01", parse(text="10^-2 - 10^-4"), parse(text="10^-4 - 10^-6"), parse(text="10^-6 - 10^-8"), paste("< ", parse(text="10^-8")))

p = stratifiedOddsRatioPlot(or.df, "Minimum p value (Katie IBD GWAS)", "Pagerank\npercentile") +
  ggtitle("Curated seed gene network - Katie's GWAS minp") +
  theme_bw(12) +
  theme(axis.text.x = element_text(angle = 25, hjust = 1), axis.title.x = element_blank()) +
  scale_x_discrete(breaks = minp_cat_levels, labels = minp_cat_display_levels)
print(p)

if (saveFiles) {
  pdf(file.path(analysis_dir, "minp_enrichment.curated_network.pdf"), width=5.5, height=4)
  print(p + theme(plot.title = element_blank()) + xlab("Minimum p value"))
  dev.off()
}
```

We can look at the same thing for gene minp values for Jimmy's GWAS.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
network1.jimmy.minp.df$minp_cat = categorize(network1.jimmy.minp.df$minp, thresholds = c(0.01, 1e-4, 1e-6, 1e-8))
minp_cat_levels = c("> 0.01", "1e-04-0.01", "1e-06-1e-04", "1e-08-1e-06", "<= 1e-08")
network1.jimmy.minp.df$minp_cat = factor(network1.jimmy.minp.df$minp_cat, levels = minp_cat_levels)

network1.jimmy.minp.df$pagerank_pctile.cat = categorize(network1.jimmy.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90))
pagerank_pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "> 90")
network1.jimmy.minp.df$pagerank_pctile.cat = categorize(network1.jimmy.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90, 95, 98))
pagerank_pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "95-98", "> 98")
network1.jimmy.minp.df$pagerank_pctile.cat = factor(network1.jimmy.minp.df$pagerank_pctile.cat, levels = pagerank_pct_cat_levels)
#xtabs(~ minp_cat + pagerank_pctile.cat, data = network1.jimmy.minp.df)
or.df = stratifiedOddsRatios(network1.jimmy.minp.df, "minp_cat", "pagerank_pctile.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = pagerank_pct_cat_levels)
stratifiedOddsRatioPlot(or.df, "Minimum p value (Jimmy IBD GWAS)", "Pagerank\npercentile") + ggtitle("Curated seed gene network - Jimmy's GWAS minp")

```

The results are very similar, so we only use Katie's GWAS in future plots.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
network2.minp.df = network2.summary.df %>%
  inner_join(genes.minp.df, by="gene_id")

network2.minp.df$minp_cat = categorize(network2.minp.df$minp, thresholds = c(0.01, 1e-4, 1e-6, 1e-8))
minp_cat_levels = c("> 0.01", "1e-04-0.01", "1e-06-1e-04", "1e-08-1e-06", "<= 1e-08")
network2.minp.df$minp_cat = factor(network2.minp.df$minp_cat, levels = minp_cat_levels)

network2.minp.df$pagerank_pctile.cat = categorize(network2.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90))
pagerank_pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "> 90")
network2.minp.df$pagerank_pctile.cat = categorize(network2.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90, 95, 98))
pagerank_pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "95-98", "> 98")
network2.minp.df$pagerank_pctile.cat = factor(network2.minp.df$pagerank_pctile.cat, levels = pagerank_pct_cat_levels)
#xtabs(~ minp_cat + pagerank_pctile.cat, data = network2.minp.df)
or.df = stratifiedOddsRatios(network2.minp.df, "minp_cat", "pagerank_pctile.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = pagerank_pct_cat_levels)
stratifiedOddsRatioPlot(or.df, "Minimum p value", "Pagerank\npercentile") + ggtitle("L2G seed gene network - Katie's GWAS minp")

```



## IBD p-value network enrichment (gene-overlapping)

We can do the same odds ratio analyses considering only variants that overlap genes, without a 10 kb window. This avoids the problem that +/-10 kb will include the same SNPs multiple times in gene-dense regions.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=6, fig.height=4}
gene_minp_file = file.path(root_dir, "genes.pc.overlapping.katie_ibd.minp.tsv.gz")

if (!file.exists(gene_minp_file)) {
  gene_pvalues_file = file.path(root_dir, "genes.pc.katie_ibd.snp_overlaps.tsv.gz")
  genes.minp.df = readr::read_tsv(gene_pvalues_file) %>%
    filter(pvalue != -1) %>%
    na.omit() %>%
    group_by(gene_id) %>%
    summarise(chr = first(chr),
              gene_window_start = first(gene_window_start),
              gene_window_end = first(gene_window_end),
              snp_pos = first(snp_pos),
              snp_id = first(snp_id),
              minp = min(pvalue)) %>%
    filter(minp > -1) %>%
    ungroup() %>%
    mutate(minp.quantile = vecQuantiles(-minp))
  write_tsv(genes.minp.df, gene_minp_file)
} 
genes.minp.df = readr::read_tsv(gene_minp_file)

network1.minp.df = network1.summary.df %>%
  inner_join(genes.minp.df, by="gene_id")

ggplot(network1.minp.df %>% filter(minp > 1e-8), aes(x=-log10(minp))) +
  geom_histogram(bins=100) + ggtitle("Distribution of minp values (Katie IBD GWAS, gene-overlapping)")
#sum(is.na(network1.minp.df$minp))
#sum(network1.minp.df$minp < 5e-8, na.rm=T)
#quantile(network1.minp.df$minp, probs=seq(0, 1, 0.01), na.rm=T)

```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
# network1.minp.df$pagerank_pctile.cat = categorize(network1.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90, 95))
# network1.minp.df$pagerank_pctile.cat = factor(network1.minp.df$pagerank_pctile.cat, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))
# 
# ggplot(network1.minp.df, aes(x=pagerank_pctile.cat, y=minp.quantile)) +
#   geom_boxplot() + geom_jitter(width=0.25, alpha=0.2)

network1.minp.df$minp_cat = categorize(network1.minp.df$minp, thresholds = c(0.01, 1e-4, 1e-6, 1e-8))
minp_cat_levels = c("> 0.01", "1e-04-0.01", "1e-06-1e-04", "1e-08-1e-06", "<= 1e-08")
network1.minp.df$minp_cat = factor(network1.minp.df$minp_cat, levels = minp_cat_levels)

network1.minp.df$pagerank_pctile.cat = categorize(network1.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90))
pagerank_pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "> 90")
network1.minp.df$pagerank_pctile.cat = categorize(network1.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90, 95, 98))
pagerank_pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "95-98", "> 98")
network1.minp.df$pagerank_pctile.cat = factor(network1.minp.df$pagerank_pctile.cat, levels = pagerank_pct_cat_levels)
#xtabs(~ minp_cat + pagerank_pctile.cat, data = network1.minp.df)
or.df = stratifiedOddsRatios(network1.minp.df, "minp_cat", "pagerank_pctile.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = pagerank_pct_cat_levels)
stratifiedOddsRatioPlot(or.df, "Minimum p value", "Pagerank\npercentile") + ggtitle("Curated seed gene network")

```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
network2.minp.df = network2.summary.df %>%
  inner_join(genes.minp.df, by="gene_id")

network2.minp.df$minp_cat = categorize(network2.minp.df$minp, thresholds = c(0.01, 1e-4, 1e-6, 1e-8))
minp_cat_levels = c("> 0.01", "1e-04-0.01", "1e-06-1e-04", "1e-08-1e-06", "<= 1e-08")
network2.minp.df$minp_cat = factor(network2.minp.df$minp_cat, levels = minp_cat_levels)

network2.minp.df$pagerank_pctile.cat = categorize(network2.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90))
pagerank_pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "> 90")
network2.minp.df$pagerank_pctile.cat = categorize(network2.minp.df$pagerank_pctile, thresholds = c(50, 60, 70, 80, 90, 95, 98))
pagerank_pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "95-98", "> 98")
network2.minp.df$pagerank_pctile.cat = factor(network2.minp.df$pagerank_pctile.cat, levels = pagerank_pct_cat_levels)
#xtabs(~ minp_cat + pagerank_pctile.cat, data = network2.minp.df)
or.df = stratifiedOddsRatios(network2.minp.df, "minp_cat", "pagerank_pctile.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = pagerank_pct_cat_levels)
stratifiedOddsRatioPlot(or.df, "Minimum p value", "Pagerank\npercentile") + ggtitle("Strict seed gene network")

```



```{r, warning=FALSE, message=FALSE, echo=FALSE}
# PRIORITIZING GENES
# Here we write out a table with genes prioritised both by L2G and network
# First we need to find all genes within a window at each locus
windowSize = 5e5
l2g.loci.df2 = l2g.loci.df %>%
  dplyr::select(chr, l2g_mean_pos, l2g_locus_merged) %>%
  mutate(l2g_start = l2g_mean_pos - windowSize, l2g_end = l2g_mean_pos + windowSize)
l2g.gr = makeGRangesFromDataFrame(l2g.loci.df2, keep.extra.columns = T, ignore.strand = T,
                                  seqinfo = NULL, seqnames.field = "chr", start.field = "l2g_start", end.field = "l2g_end")

grch38.nodup = grch38 %>% filter(!duplicated(ensgene))
grch38.df = grch38.nodup %>%
  select(chr, gene_start = start, gene_end = end, ensgene, symbol)
grch38.gr = makeGRangesFromDataFrame(grch38.df, keep.extra.columns = T, ignore.strand = T,
                                             seqinfo = NULL, seqnames.field = "chr", start.field = "gene_start", end.field = "gene_end")

hits.df = as_tibble( findOverlaps(l2g.gr, grch38.gr, ignore.strand = T) )

# Get all genes within the window around or L2G loci, and then merge in
# each of the network pageranks
locus_genes.df = bind_cols(l2g.loci.df[hits.df$queryHits,] %>%
                             select(l2g_locus_merged, chr, l2g_mean_pos, top_gene_mean, top_gene_score_mean, lead_p),
                           grch38.df[hits.df$subjectHits,] %>% dplyr::select(-chr) %>% dplyr::rename(gene_id = ensgene)) %>%
  left_join(l2g.genes.df %>% select(l2g_locus_merged, gene_id, l2g_score_mean), by=c("gene_id", "l2g_locus_merged")) %>%
  left_join(network1.summary.df %>% select(gene_id, geneSymbol, network1_pagerank_pctile = pagerank_pctile, network1_seed_gene = seed_gene)) %>%
  left_join(network2.summary.df %>% select(gene_id, network2_pagerank_pctile = pagerank_pctile, network2_seed_gene = seed_gene))

# Create a network score that rewards being above the 50th pctile and is the
# average across the 3 networks
# Then get ranks of genes at each locus based on their network scores
locus_genes.df = locus_genes.df %>%
  group_by(l2g_locus_merged) %>%
  mutate(network1_rank = rank(-network1_pagerank_pctile, ties.method = "average"),
         network2_rank = rank(-network2_pagerank_pctile, ties.method = "average")) %>%
  rowwise() %>%
  mutate(network_avg = mean(c(network1_pagerank_pctile, network2_pagerank_pctile)),
         network_max = max(c(network1_pagerank_pctile, network2_pagerank_pctile)),
         network_score = max(0, (network_max - 50) * 2),
         l2g_network_score = sqrt(l2g_score_mean) * network_score) %>%
  group_by(l2g_locus_merged) %>%
  mutate(l2g_network_score_rank = rank(-l2g_network_score, ties.method = "average")) %>%
  arrange(chr, l2g_mean_pos, l2g_locus_merged, desc(l2g_network_score))
write_tsv(locus_genes.df, file.path(analysis_dir, "locus_genes.network.tsv"), na = "")

#ViewDups(locus_genes.df, "gene_id")
```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
# Flag genes where the gene has a higher network score than the top L2G gene
# and also has a L2G score of at least 0.1
locus_top_l2gene_network_score = locus_genes.df %>%
  group_by(l2g_locus_merged) %>%
  arrange(desc(l2g_score_mean)) %>%
  mutate(locus_has_curated_gene = any(network1_seed_gene, na.rm = T)) %>%
  filter(row_number() == 1) %>%
  select(l2g_locus_merged, top_gene_network_score = network_score, locus_has_curated_gene) %>%
  mutate(top_gene_network_score = if_else(is.na(top_gene_network_score), 0, top_gene_network_score))
locus_second_best_network_candidate = locus_genes.df %>%
  filter(network_avg > 80, l2g_score_mean > 0.1) %>%
  group_by(l2g_locus_merged) %>%
  arrange(desc(network_score)) %>%
  filter(row_number() == 2) %>%
  select(l2g_locus_merged, second_best_network_score = network_score)
locus_second_best_l2g_net_candidate = locus_genes.df %>%
  group_by(l2g_locus_merged) %>%
  arrange(desc(l2g_network_score)) %>%
  filter(row_number() == 2) %>%
  select(l2g_locus_merged, second_best_l2g_network_score = l2g_network_score)
locus_genes.df2 = locus_genes.df %>%
  group_by(gene_id) %>%
  left_join(locus_top_l2gene_network_score, by="l2g_locus_merged") %>%
  left_join(locus_second_best_network_candidate, by="l2g_locus_merged") %>%
  left_join(locus_second_best_l2g_net_candidate, by="l2g_locus_merged") %>%
  mutate(second_best_network_score = if_else(is.na(second_best_network_score), 0, second_best_network_score)) %>%
  mutate(network_revised_candidate = (!locus_has_curated_gene & network_score > top_gene_network_score & network_score > 60 & top_gene_network_score < 60 & l2g_score_mean > 0.1 & l2g_network_score_rank == 1),
         network_top_new_candidate = (!locus_has_curated_gene & network_score > 90 & l2g_score_mean > 0.1 & l2g_network_score_rank == 1 & second_best_network_score < 60),
         network_l2g_top_new_candidate = (!locus_has_curated_gene & l2g_network_score_rank == 1 & (l2g_network_score - second_best_l2g_network_score) > 20)) %>%
  left_join(grch38.nodup %>% select(ensgene, description), by=c("gene_id" = "ensgene")) %>%
  arrange(chr, l2g_mean_pos, l2g_locus_merged, desc(l2g_score_mean)) %>%
  select(l2g_locus_merged:gene_id, network_top_new_candidate, network_l2g_top_new_candidate, symbol, l2g_network_score, l2g_score_mean, network1_pagerank_pctile, network2_pagerank_pctile, network1_seed_gene, network2_seed_gene, second_best_l2g_network_score, second_best_network_score, l2g_network_score_rank, everything())
#sum(locus_genes.df2$network_revised_candidate, na.rm=T)
sum(locus_genes.df2$network_revised_candidate, na.rm=T)
sum(locus_genes.df2$network_top_new_candidate, na.rm=T)
sum(locus_genes.df2$network_l2g_top_new_candidate, na.rm=T)
write_tsv(locus_genes.df2, file.path(analysis_dir, "locus_genes.network.revised_candidates.tsv"), na = "")
```


## Sub-threshold genes

Let's look for genes with high network scores that have sub-genome-wide significant p values. This is the distribution of network scores after excluding genome-wide significant loci. (But actually I need a better way to do this, because I've only removed genes with a "locus2gene" score - and some genes remain at the GWAS loci, which suggests that we're missing some L2G scores. This could be due to LD and the window size used for L2G.)

```{r, warning=FALSE, message=FALSE, echo=FALSE}
network1.subthreshold.df = network1.minp.df %>%
  filter(!gene_id %in% l2g.genes.df$gene_id)

View(network1.subthreshold.df %>% filter(minp > 5e-8) %>% arrange(minp) %>% select(geneSymbol, degree, pagerank_pctile, minp, chr, gene_window_start, gene_window_end, everything()))

ggplot(network1.minp.df, aes(x=minp_cat, y=pagerank_pctile)) +
  geom_violin(alpha = 0.9) + geom_jitter(width = 0.25, alpha = 0.2) + geom_boxplot(width = 0.05, outlier.shape = NA, fill="white")


```


